### ISSUE TITLE
Results change during testing / evaluation

### ISSUE THREAD

**Issue by yt2639** [Pointcept/Pointcept#508]
Hii authors, thanks for providing the code and it is greatly helpful. I observed an issue during testing / evaluation of PTv3. An example code is shown below:
```
from PTv3.code.pointcept.models.point_prompt_training.point_prompt_training_v1m2_decoupled import PointPromptTraining
from PTv3.code.pointcept.models.point_transformer_v3.point_transformer_v3m1_base import PointTransformerV3

backbone=dict(
                        in_channels=15, # originally this is 6, but I need to use a different type of input so input dim is now 15
                        order=('z', 'z-trans', 'hilbert', 'hilbert-trans'),
                        stride=(2, 2, 2, 2),
                        enc_depths=(3, 3, 3, 6, 3),
                        enc_channels=(48, 96, 192, 384, 512),
                        enc_num_head=(3, 6, 12, 24, 32),
                        enc_patch_size=(1024, 1024, 1024, 1024, 1024),
                        dec_depths=(3, 3, 3, 3),
                        dec_channels=(64, 96, 192, 384),
                        dec_num_head=(4, 6, 12, 24),
                        dec_patch_size=(1024, 1024, 1024, 1024),
                        mlp_ratio=4,
                        qkv_bias=True,
                        qk_scale=None,
                        attn_drop=0.0,
                        proj_drop=0.0,
                        drop_path=0.3,
                        shuffle_orders=True,
                        pre_norm=True,
                        enable_rpe=False,
                        enable_flash=True,
                        upcast_attention=False,
                        upcast_softmax=False,
                        cls_mode=False,
                        pdnorm_bn=True,
                        pdnorm_ln=True,
                        pdnorm_decouple=True,
                        pdnorm_adaptive=False,
                        pdnorm_affine=True,
                        pdnorm_conditions=('ScanNet', 'ScanNet200', 'ScanNet++',
                                        'Structured3D', 'ALC'))

backbone = PointTransformerV3(**backbone)

model = dict(
                    backbone=backbone,
                    context_channels=256,
                    conditions=('ScanNet', 'ScanNet200', 'ScanNet++', 'Structured3D', 'ALC'),
                    num_classes=(20, 200, 100, 25, 185))
        
self.segmentor = PointPromptTraining(**model)
```
Then, I obtain the results using
```
ptv3_output = self.segmentor(data_dict) # [total_num_points, 64]
```
I made sure all modules `self.training` is False and all parameters `requires_grad` is False. Also, I made sure `data_dict` is not changed. Then, when I run the above code several times, it gives totally different results which is definitely not due to floating precision sort of stuff.

I am wondering what I was missing here? How do we obtain a deterministic result during testing / evaluation?

Many thanks!

**Comment by yt2639** [comment]
I actually changed `PointPromptTraining` to be a feature extractor only if that matters. The details are below:
```
class PointPromptTraining(nn.Module):
    """
    PointPromptTraining v1m2 provides Data-driven Context and enables multi-dataset training with
    Decoupled Segmentation Head. PDNorm is supported by SpUNet-v1m3 to adapt the
    backbone to a specific dataset with a given dataset condition and context.
    """

    def __init__(
            self,
            backbone=None,
            context_channels=256,
            conditions=('ScanNet', 'ScanNet200', 'ScanNet++', 'Structured3D','ALC'),
            num_classes=(20, 200, 100, 25, 185),
    ):
        super().__init__()
        assert len(conditions) == len(num_classes)
        self.backbone = backbone
        self.conditions = conditions
        self.embedding_table = nn.Embedding(len(conditions), context_channels)

    def forward(self, data_dict):
        condition = 'ScanNet200'  # data_dict["condition"][0]
        assert condition in self.conditions
        context = self.embedding_table(
            torch.tensor([self.conditions.index(condition)],
                         device=data_dict["coord"].device))
        data_dict["context"] = context
        point = self.backbone(data_dict) # [total_num_points, backbone_out_channels = 64]

        if isinstance(point, Point):
            feat = point.feat
        else:
            feat = point
        
        return feat
```

**Comment by Gofinge** [comment]
Hey, there are some operators like `spconv`, `torch_scatter`, that are non-deterministic in CUDA. A slight difference caused by non-deterministic operators won't be a problem in actual usage.

**Comment by yt2639** [comment]
Hii @Gofinge Thanks for your response and I really appreciate it. Actually the difference is not that small even if after disabling `shuffle_orders` (i.e., `shuffle_orders = False`). Here is an example of the output given the exactly same input:
```
# first time
tensor([[ 0.0731,  0.2487,  0.2329,  ...,  0.0938, -0.0612,  0.4966],
        [-0.1252,  0.1680,  0.1249,  ...,  0.0655, -0.2164,  0.2825],
        [ 0.0202,  0.3110,  0.0163,  ...,  0.2063, -0.1097,  0.1053],
        ...,
        [ 0.0342,  0.1149, -0.0064,  ...,  0.0739,  0.0105,  0.0740],
        [ 0.0346,  0.0366,  0.0398,  ...,  0.0277,  0.0458,  0.0514],
        [ 0.0665,  0.1172,  0.1082,  ...,  0.0213,  0.0249,  0.0519]],
       dtype=torch.float16)

# second time
tensor([[-0.1399,  0.1526,  0.1511,  ...,  0.0786, -0.2141,  0.3057],
        [-0.0063,  0.3054,  0.0287,  ...,  0.2250, -0.0693,  0.1017],
        [-0.0987,  0.4573,  0.3213,  ...,  0.1661,  0.0804,  0.2253],
        ...,
        [ 0.0183,  0.0934,  0.1033,  ...,  0.0473,  0.0607,  0.0569],
        [ 0.0778,  0.1451,  0.0120,  ...,  0.0652, -0.0208,  0.0867],
        [ 0.0070,  0.1125, -0.0131,  ...,  0.0765,  0.0211,  0.0663]],
       dtype=torch.float16)
```
I also tried with `float32` but it appears to have similar behavior as the example above.

I am trying to use PTv3 in my other projects and the non-deterministic behavior is causing severe overfitting to my model. Do you have any idea that how we can make everything deterministic? Thanks!

**Comment by Gofinge** [comment]
Hey, I think it's normal,
1. If you mean the difference between eval and test, the evaluation/testing pipeline is a bit different; the test process is a bit slow but more precise.
2. If you mean repeating the test process, because some 3D operator is not deterministic , I think it is also not necessary to make it deterministic, so there would be slightly different for each forward, but won't affect number metric.
